Add case test_uninstall

chriscchien opened the pull request:
Add case test_uninstall

longhorn/longhorn#4031

Signed-off-by: Chris Chien <chris.chien@suse.com>

innobead commented on test_framework/scripts/longhorn-uninstall.sh:
```diff
@@ -0,0 +1,69 @@
+  LONGHORN_UNINSTALL_TEST_POD_NAME="longhorn-test-uninstall" 
+  run_uninstall_longhorn_test
+
+  mv ${TF_VAR_tf_workspace}/longhorn-test-junit-report.xml ${TF_VAR_tf_workspace}/longhorn-test-junit-report-temp.xml
+  jrm ${TF_VAR_tf_workspace}/longhorn-test-junit-report.xml "${TF_VAR_tf_workspace}/*.xml"
```
You can just follow the current way we have for aggregate the report by junit Jenkins plugin. However, I like you use the dependencies in the container image, but we have the pattern already, so I would suggest just following the existing one.

https://plugins.jenkins.io/junit/
```
junit `*.xml`
```

or check an example in test_framework/Jenkinsfile
```
summary = junit 'longhorn-test-upgrade-from-stable-junit-report.xml, longhorn-test-upgrade-from-transient-junit-report.xml, longhorn-test-junit-report.xml'
```

innobead commented on test_framework/Dockerfile.setup:
```diff
@@ -30,7 +30,9 @@ RUN wget -q https://storage.googleapis.com/kubernetes-release/release/$KUBECTL_V
-    ./get_helm.sh
-
+    ./get_helm.sh && \
+    apk add --update nodejs npm && \
+    apk add --update npm && \
+    npm install -g junit-report-merger
```
Remove, because use junit Jenkins plugin instead.

innobead commented on manager/integration/tests/test_uninstall.py:
```diff
@@ -0,0 +1,60 @@
+
+    longhorn_version = os.getenv("LONGHORN_REPO_BRANCH")
+
+    if longhorn_version == "master" or longhorn_version == "v1.4.x":
+        delete_confirmtion_flag = \
+            client.by_id_setting(SETTING_DELETING_CONFIRMATION_FLAG)
+        setting = client.update(delete_confirmtion_flag, value="true")
+        assert setting.name == SETTING_DELETING_CONFIRMATION_FLAG
+        assert setting.value == "true"
```
You can just try to see if the setting exists to determine if the testing version has to the deletion flag support instead of hard-coding the version check.

innobead commented on manager/integration/tests/test_uninstall.py:
```diff
@@ -0,0 +1,60 @@
+    client = get_longhorn_api_client()
+    lh_raw = "https://raw.githubusercontent.com/longhorn/longhorn"
+
+    longhorn_version = os.getenv("LONGHORN_REPO_BRANCH")
```
Why not just prepare the manifest path from outside? like https://github.com/longhorn/longhorn-tests/blob/f47c2879d5dff13311d71f9312226e8da14a1690/test_framework/scripts/longhorn-setup.sh#L18-L18. Try to follow the same pattern.

innobead commented on manager/integration/tests/test_uninstall.py:
```diff
@@ -0,0 +1,60 @@
+    command = "kubectl create -f {}/{}/uninstall/uninstall.yaml".\
+              format(lh_raw, longhorn_version)
+
+    os.system(command)
```
nit: it's nice to check error or return value.

innobead commented on manager/integration/tests/test_uninstall.py:
```diff
@@ -0,0 +1,60 @@
+        time.sleep(RETRY_INTERVAL_LONG)
+    assert uninstall_job_ready is True
+
+    command = "kubectl delete -f {}/{}/deploy/longhorn.yaml"\
```
There is an injected manifest env variable, so why redo this? ex: LONGHORN_MANIFEST_URL

innobead commented on manager/integration/tests/test_uninstall.py:
```diff
@@ -0,0 +1,60 @@
+              .format(lh_raw, longhorn_version)
+    os.system(command)
+
+    command = "kubectl delete -f {}/{}/uninstall/uninstall.yaml"\
```
The same.

innobead commented on manager/integration/tests/test_uninstall.py:
```diff
@@ -0,0 +1,60 @@
+    longhorn_ns_deleted = False
+    for i in range(RETRY_COUNTS):
+        namespaces = core_api.list_namespace()
+        find = False
```
`find` is too general. Having a good habit to make the variable name meaningful. Also, why not just using one variable instead of two in this case?

```python
longhorn_deleted = False

for i in range(RETRY_COUNTS):
    namespaces = core_api.list_namespace()

    for namespace in namespaces.items:
        if namespace.metadata.name == "longhorn-system":
            longhorn_deleted = False
            break

    if not longhorn_deleted:
        time.sleep(RETRY_INTERVAL_LONG)

assert longhorn_deleted
```

innobead reviewed (changes requested):
In general, the concept is OK, but some comments needed to resolve.

innobead commented:
@yangchiu Please review this.

yangchiu reviewed (commented):
Is it possible to follow the upgrade test pattern like https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/scripts/upgrade-longhorn.sh ?
So we don't have to write Python code to do the uninstall, and the `uninstall-longhorn.sh` will be a standalone plain shell script having no dependencies on our test code.

innobead commented:
> Is it possible to follow the upgrade test pattern like https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/scripts/upgrade-longhorn.sh ? So we don't have to write Python code to do the uninstall, and the `uninstall-longhorn.sh` will be a standalone plain shell script having no dependencies on our test code.

sounds good to me

chriscchien commented:
Updated PR content for review, thank you

yangchiu commented:
What's the difference between [uninstall-longhorn.sh](https://github.com/longhorn/longhorn-tests/pull/1253/files#diff-22f8782fedf9d78c798964b7e5f185da7a696cc26f71caabda29e464bb1a745e) and [longhorn-uninstall.sh](https://github.com/longhorn/longhorn-tests/pull/1253/files#diff-5f5271d2784628aea270fa6914367e3c0ee87beac22a43daebc2389469be7155)?

chriscchien commented:
> What's the difference between [uninstall-longhorn.sh](https://github.com/longhorn/longhorn-tests/pull/1253/files#diff-22f8782fedf9d78c798964b7e5f185da7a696cc26f71caabda29e464bb1a745e) and [longhorn-uninstall.sh](https://github.com/longhorn/longhorn-tests/pull/1253/files#diff-5f5271d2784628aea270fa6914367e3c0ee87beac22a43daebc2389469be7155)?

[longhorn-uninstall.sh](https://github.com/longhorn/longhorn-tests/pull/1253/files#diff-5f5271d2784628aea270fa6914367e3c0ee87beac22a43daebc2389469be7155) : Create longhorn-test pod with specific pytest command to run uninstall test case and have junit report

[uninstall-longhorn.sh](https://github.com/longhorn/longhorn-tests/pull/1253/files#diff-22f8782fedf9d78c798964b7e5f185da7a696cc26f71caabda29e464bb1a745e) : Shell script to execute uninstall command

khushboo-rancher commented:
Can we rename the file `longhorn-uninstall.sh` to `longhorn-uninstall-setup.sh`?

chriscchien commented:
Rebase to master and here is the full [jenkins log](https://github.com/longhorn/longhorn-tests/files/11293164/jenkinsoutput.txt) from my test environment

![Screenshot from 2023-04-21 15-14-23](https://user-images.githubusercontent.com/73337386/233567275-c27d44df-cb6c-4b08-9ccc-fdd212903fea.png)

cc @longhorn/qa

khushboo-rancher commented:
Thanks @chriscchien, this looks good.

Just one question, is it possible to check error/warning in the logs of longhorn-uninstallation job?
Also, can we check after uninstallation, no crd is left `kubectl get crd | grep longhorn`

chriscchien commented:
Hi @khushboo-rancher, updated PR for adding longhorn CRD check at `uninstall-longhorn.sh`

![Screenshot from 2023-04-28 09-02-25](https://user-images.githubusercontent.com/73337386/235029854-4bbdd7da-00d9-46c2-aa3c-13d360ceba2b.png)

![Screenshot from 2023-04-28 09-05-11](https://user-images.githubusercontent.com/73337386/235030130-4710922a-733e-47b5-8c3f-776d7dd6ae4f.png)


[jenjinslog.txt](https://github.com/longhorn/longhorn-tests/files/11348995/jenjinslog.txt)

khushboo-rancher reviewed (approved):
@yangchiu Could you also take a look at it?

yangchiu reviewed (dismissed):
LGTM

khushboo-rancher commented on test_framework/scripts/longhorn-uninstall-setup.sh:
```diff
@@ -0,0 +1,71 @@
+
+run_uninstall_longhorn_test(){
+
+  LONGHORN_TESTS_CUSTOM_IMAGE=${LONGHORN_TESTS_CUSTOM_IMAGE:-"longhornio/longhorn-manager-test:master-head"}
```
Just one last doubt, the custom image should come from the specified branch in the jenkins param, right?

chriscchien replied:
Hi @khushboo-rancher , updated!

innobead commented on test_framework/scripts/longhorn-uninstall-setup.sh:
```diff
@@ -0,0 +1,71 @@
+  ARCH=${1}
+  BASEDIR=${2}
+
+    if [[ ${ARCH} == "amd64" ]] ; then
+    if [[ ${TF_VAR_k8s_distro_name} == [rR][kK][eE] ]]; then
+      export KUBECONFIG="${BASEDIR}/kube_config_rke.yml"
+    elif [[ ${TF_VAR_k8s_distro_name} == [rR][kK][eE]2 ]]; then
+      export KUBECONFIG="${BASEDIR}/terraform/aws/${DISTRO}/rke2.yaml"
+    else
+      export KUBECONFIG="${BASEDIR}/terraform/aws/${DISTRO}/k3s.yaml"
+    fi
+  elif [[ ${ARCH} == "arm64"  ]]; then
+    export KUBECONFIG="${BASEDIR}/terraform/aws/${DISTRO}/k3s.yaml"
+  fi
```
Please make sure the bash format correctly.

chriscchien replied:
Hi @innobead , could you give me more hint about the bash format need be updated because this part of code I referenced from longhorn-setup.sh, I will update immediately, thank you

innobead reviewed (changes requested):
Rename script/longhorn-uninstall-setup.sh to script/longhorn-uninstall.sh, the same patter as other scripts in the same folder.

longhorn-uninstall.sh only supports uninstalling by the manifest way, but actually we support different ways in longhorn-install/setup.sh like helm cahrt.

I would suggest having all uninstall methods the same as install/setup.

chriscchien commented:
> In general, LGTM partially.
> 
> Below items needs to fix.
> 
> * Rename script/longhorn-uninstall-setup.sh to script/longhorn-uninstall.sh, the same pattern as other scripts in the same folder.
> * longhorn-uninstall.sh only supports uninstalling by the manifest way, but actually, we support different ways in longhorn-install/setup.sh such as helm chart. Have all uninstall methods the same as install/setup. (optional)
> 
> If the second item we can not achieve in a short time, just make sure only run uninstall testing when the installation is via manifest.

Renamed file to `scripts/longhorn-uninstall.sh` and modified `test_framework/Jenkinsfile` to execute uninstall case when Longhorn is installed by default method. For helm chart uninstall cases, created this [ticket](https://github.com/longhorn/longhorn/issues/5938)  for tracking, I will update this PR if I missed anything, thank you.

chriscchien commented:
Conflict fixed

innobead commented:
@yangchiu review this. thanks.

yangchiu reviewed (approved):
LGTM

innobead reviewed (dismissed):
In general, LGTM.

Some minor things can be updated if needed.

chriscchien commented on manager/integration/tests/test_uninstall.py:
```diff
@@ -0,0 +1,57 @@
+@pytest.mark.uninstall   # NOQA
+def test_uninstall(longhorn_uninstall_manifest_url,longhorn_deploy_manifest_url): # NOQA
+
+    client = get_longhorn_api_client()
```
Fixture will run teardown after test case complete but at that moment LH uninstalled, `cleanup_client()` will fail make test case fail too.

chriscchien commented on manager/integration/tests/conftest.py:
```diff
@@ -96,6 +103,20 @@ def pytest_addoption(parser):
+                     will be used in test_uninstall
+                     (default: https://raw.githubusercontent.com/longhorn/longhorn/master/uninstall/uninstall.yaml)''') # NOQA
+
+    parser.addoption(DEPLOY_LH_MANIFEST_URL, action="store",
```
We need to run `kubectl delete -f https://raw.githubusercontent.com/longhorn/longhorn/<branch>/deploy/longhorn.yaml` when uninstall Longhorn. Currently the steps executed in pytest test case and calling `uninstall-longhorn.sh`, use `DEPLOY_LH_MANIFEST_URL` as a parameter, I will update if need to modify, thank you.

chriscchien commented:
Updated and rebase to master head, below is my local env test result, thank you
![image](https://github.com/longhorn/longhorn-tests/assets/73337386/98c4b1e5-907d-445e-842f-c837fa4b67ac)

yangchiu commented:
LGTM. Suggest merge this into `master` first to make it actually run and see if there's anything can be improved.
